### PR TITLE
GenerativeClassifier Model

### PR DESCRIPTION
**Pull request by sheldon-roberts**
This pull request introduces `GenerativeClassifier`, a new model that treats text classification as a sequence‑generation task.  Instead of attaching a classification head on top of a transformer embedding, we fine‑tune a causal language model to generate the ground‑truth label string.

Key Benefits:
1. **Supports Extreme Classification** - The model can output an arbitrary number of labels without changing the final layer size.
2. **Zero/Few‑Shot Performance** - Leverage prompting with the pre-trained model to achieve strong few-shot performance.
3. **Constrained Decoding** – A prefix‑trie keeps generation strictly inside the allowed label set.
4. **Semantic label reasoning** – Label names are natural text, letting the model reason about their meaning and similarity.

## Usage example
```python
from transformers import AutoModelForCausalLM, AutoTokenizer
from flair.datasets import GO_EMOTIONS
from flair.trainers import ModelTrainer
from flair.data import Sentence
from flair.models import GenerativeClassifier

corpus = GO_EMOTIONS().downsample(0.01)
LABEL_TYPE = "emotion"
MODEL_NAME = "Qwen/Qwen2.5-0.5B"

label_dict = corpus.make_label_dictionary(label_type=LABEL_TYPE, add_unk=False)
tok = AutoTokenizer.from_pretrained(MODEL_NAME)
lm = AutoModelForCausalLM.from_pretrained(MODEL_NAME)

classifier = GenerativeClassifier(
    causal_model=lm,
    tokenizer=tok,
    label_dictionary=label_dict,
    label_type=LABEL_TYPE,
    prompt_template=("<Task> Classify the emotion of the following text </Task> Text: {text}  \nEmotions: "),
)

trainer = ModelTrainer(classifier, corpus)
trainer.fine_tune(
    base_path="resources/taggers/gencls_goemotions",
    mini_batch_size=8,
    max_epochs=5,
)

s = Sentence("I cant wait to try it! It looks so good")
classifier.predict(s)
print(s)
```

### CODE REVIEW

File: flair/training_utils.py [hunk 1]
```diff
@@ -524,3 +527,95 @@ def create_labeled_sentence_from_entity_offsets(
         token_entities = [entity for entity in token_entities if entity.end_token_idx <= token_limit]
 
     return create_labeled_sentence_from_tokens(tokens, token_entities)
+
+
+def load_model_components(
```
Comment (sheldon-roberts) [inline flair/training_utils.py]: (edited): You need to use this `load_and_tie_causal_model` helper function to train a generative model in a multitask setup with other model types (like `TextClassifier` or `SpanClassifier`).

I wasn't sure what the best place to put this function is. I've added it to training utils for now.

Here is an example of how it's used to train MTL:

```python
from flair.data import Sentence
from flair.datasets import GO_EMOTIONS, TREC_6
from flair.models import TextClassifier, GenerativeClassifier
from flair.training_utils import load_and_tie_causal_model
from flair.nn.multitask import make_multitask_model_and_corpus
from flair.trainers import ModelTrainer
from transformers import AutoTokenizer
from flair.embeddings import TransformerDocumentEmbeddings

MODEL_NAME = "Qwen/Qwen2.5-0.5B" 
TRUST_REMOTE_CODE = True
DOWNSAMPLE_RATE = 0.1

embeddings = TransformerDocumentEmbeddings(MODEL_NAME)
causal_model = load_and_tie_causal_model(
    model_name=MODEL_NAME,
    embeddings=embeddings,
    trust_remote_code=TRUST_REMOTE_CODE,
)
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME, trust_remote_code=TRUST_REMOTE_CODE)

# Task 1: GoEmotions
LABEL_TYPE_1 = "emotion"
corpus_1 = GO_EMOTIONS().downsample(DOWNSAMPLE_RATE)
label_dict_1 = corpus_1.make_label_dictionary(label_type=LABEL_TYPE_1, add_unk=False)

model_1 = GenerativeClassifier(
    causal_model=causal_model,
    tokenizer=tokenizer,
    label_dictionary=label_dict_1,
    label_type=LABEL_TYPE_1,
    prompt_template=("<Task> Classify emotions </Task> Text: {text} \nEmotions: "),
    trust_remote_code=TRUST_REMOTE_CODE,
)

# Task 2: TREC Question Classification
LABEL_TYPE_2 = "question_class"
corpus_2 = TREC_6().downsample(DOWNSAMPLE_RATE)
label_dict_2 = corpus_2.make_label_dictionary(label_type=LABEL_TYPE_2)

model_2 = TextClassifier(
    embeddings=embeddings,
    label_dictionary=label_dict_2,
    label_type=LABEL_TYPE_2,
)

multitask_model, multicorpus = make_multitask_model_and_corpus(
    [
        (model_1, corpus_1),
        (model_2, corpus_2),
    ]
)
trainer = ModelTrainer(multitask_model, multicorpus)
trainer.fine_tune(
    base_path="resources/taggers/multitask_gencls_example",
    mini_batch_size=4,
    max_epochs=2,
    use_amp=True
)

print("\n--- Prediction Example (GoEmotions) ---")
s1 = Sentence("I am so excited about the new movie release!")
model_1.predict(s1)
print(s1)

print("\n--- Prediction Example (TREC) ---")
s2 = Sentence("What is the capital of Canada?")
model_2.predict(s2)
print(s2)
```

File: flair/models/generative_classification_model.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+        node.is_end = True
+
+
+def build_label_trie(tokenizer: AutoTokenizer, vocab: List[str], sep: str) -> TrieNode:
+    """Builds a TrieNode from a list of label strings."""
+    trie = TrieNode()
+    for label in vocab:
+        label = label.strip()
+        if not label:
+            continue
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: Should we perhaps warn here? Should be weird to have a label that is only whitespace

File: flair/models/generative_classification_model.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+
+
+def make_prefix_allowed_tokens_fn(
+    trie: TrieNode, tokenizer: AutoTokenizer, prompt_lens: List[int], sep_id: int
+) -> Callable[[int, torch.LongTensor], List[int]]:
+    """Factory that returns a prefix_allowed_tokens_fn for generate()."""
+    eos_id = tokenizer.eos_token_id
+    pad_id = tokenizer.pad_token_id
+
+    def walk(node: TrieNode, tokens: List[int]) -> Optional[TrieNode]:
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: Maybe call `tokens` -> `token_ids` and might make sense to make this a member of the `Trie` class. quick docstring would help

File: flair/models/generative_classification_model.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+        for t in tokens:
+            node = node.children.get(t)
+            if node is None:
+                return None
+        return node
+
+    def prefix_fn(batch_id: int, full_ids: torch.LongTensor) -> List[int]:
+        # remove prompt and padding tokens
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: Docstring here pls, not clear at first glance what this function does. It seems like this is what ensures that you specifically output entire labels

File: flair/models/generative_classification_model.py [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+        if not label:
+            continue
+        token_ids = tokenizer.encode(f" {label}", add_special_tokens=False)
+        if token_ids:
+            trie.insert(token_ids)
+    return trie
+
+
+def make_prefix_allowed_tokens_fn(
+    trie: TrieNode, tokenizer: AutoTokenizer, prompt_lens: List[int], sep_id: int
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: `prompt_lengths` might be a better name here, when I see lens I think optical

File: flair/models/generative_classification_model.py [hunk 6]
```diff
@@ -0,0 +1,457 @@
...
+            )
+        sep_id = ids[0]
+    return sep_id
+
+
+@dataclass
+class TrieNode:
+    """Trie for fast prefix-based filtering of label token sequences."""
+
+    children: Dict[int, "TrieNode"] = field(default_factory=dict)
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: Don't need to change this, but with the version of Python Flair is on now, we can just use `dict` and the likes and avoid imports

File: flair/models/generative_classification_model.py [hunk 7]
```diff
@@ -0,0 +1,457 @@
...
+        token_ids = tokenizer.encode(f" {label}", add_special_tokens=False)
+        if token_ids:
+            trie.insert(token_ids)
+    return trie
+
+
+def make_prefix_allowed_tokens_fn(
+    trie: TrieNode, tokenizer: AutoTokenizer, prompt_lens: List[int], sep_id: int
+) -> Callable[[int, torch.LongTensor], List[int]]:
+    """Factory that returns a prefix_allowed_tokens_fn for generate()."""
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: I think it's worth defining the inputs here since it's not obvious

File: flair/models/generative_classification_model.py [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+        ids = full_ids.view(-1).tolist()
+        while ids and ids[0] == pad_id:
+            ids.pop(0)
+        ids = ids[prompt_lens[batch_id] :]
+
+        if not ids:
+            return list(trie.children.keys())
+
+        # find the start of the last segment (tokens after the last separator)
+        last_sep_indices = [i for i, x in enumerate(ids) if x == sep_id]
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: Is this just all separator indices?

File: flair/models/generative_classification_model.py [hunk 9]
```diff
@@ -0,0 +1,457 @@
...
+
+    def insert(self, token_ids: List[int]) -> None:
+        """Insert a sequence of token IDs into this trie."""
+        node = self
+        for tok in token_ids:
+            node = node.children.setdefault(tok, TrieNode())
+        node.is_end = True
+
+
+def build_label_trie(tokenizer: AutoTokenizer, vocab: List[str], sep: str) -> TrieNode:
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: Is `sep` used here?

File: flair/training_utils.py [hunk 10]
```diff
@@ -524,3 +527,95 @@ def create_labeled_sentence_from_entity_offsets(
...
+    if use_flair_embedding:
+        logger.info(f"Loading Flair TransformerEmbeddings for '{model_name}'")
+        try:
+            flair_embedding = TransformerEmbeddings(
+                model_name,
+                fine_tune=fine_tune,
+                trust_remote_code=trust_remote_code,
+                is_document_embedding=is_document_embedding,
+                is_token_embedding=is_token_embedding,
+                cls_pooling=cls_pooling,
```
Comment (MattGPT-ai) [inline flair/training_utils.py]: We might want to also allow passing of kwargs into here if this isn't an exhaustive list of params you could potentially want to pass in

File: flair/training_utils.py [hunk 11]
```diff
@@ -524,3 +527,95 @@ def create_labeled_sentence_from_entity_offsets(
...
+                    logger.warning(f"Weight tying via '{attr}' failed: {tie_err}")
+                break
+        if not tied:
+            logger.warning(
+                "Could not find any of 'base_model', 'model', or 'transformer' in causal_model; skipping weight tying."
+            )
+
+        # disable use_cache for safe multitask training
+        if hasattr(flair_embedding.model.config, "use_cache"):
+            flair_embedding.model.config.use_cache = False
```
Comment (MattGPT-ai) [inline flair/training_utils.py]: Why is this necessary here? Do you know if this slows down generation significantly?

File: flair/models/generative_classification_model.py [hunk 12]
```diff
@@ -0,0 +1,457 @@
...
+                gen_only = gen_ids[:, tok["input_ids"].shape[1] :]  # remove prompt tokens
+                outputs = self.tokenizer.batch_decode(gen_only, skip_special_tokens=True)
+
+                valid_label_set = set(self.label_dictionary.get_items()) if self.use_constrained_decoding else None
+                for sentence, out_str in zip(batch, outputs):
+                    sentence.remove_labels(label_type)
+                    generated_items = {s.strip() for s in out_str.split(self.separator) if s.strip()}
+                    for item in generated_items:
+                        if self.use_constrained_decoding:
+                            if item in valid_label_set:
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: I added a function `Dictionary.has_item` that should be faster than creating a set from `get_items()` every time

File: flair/models/generative_classification_model.py [hunk 13]
```diff
@@ -0,0 +1,457 @@
...
+            prompt = self._format_prompt(sentence.text)
+            gold_str = self._gold_string(sentence)
+
+            if not gold_str:
+                continue
+
+            g_ids = self.tokenizer.encode(gold_str, add_special_tokens=False)
+
+            max_prompt_len = max(0, self.max_length - len(g_ids) - 1)  # reserve room for gold + EOS
+            t_ids = self.tokenizer.encode(
```
Comment (MattGPT-ai) [inline flair/models/generative_classification_model.py]: Better variable names for the ids would be helpful

**Review by MattGPT-ai — COMMENTED** [review]
Looking pretty good, first round of comments here, mostly minor. Still gonna take a second look at the main parts of the model and prediction functions

File: flair/models/generative_classification_model.py (hunk 4 above)
Comment (sheldon-roberts) [reply flair/models/generative_classification_model.py]: docstring added. And yes, this function ensures that the model generates full and valid labels

File: flair/models/generative_classification_model.py (hunk 8 above)
Comment (sheldon-roberts) [reply flair/models/generative_classification_model.py]: yes it is. I renamed the variable for clarity

File: flair/models/generative_classification_model.py (hunk 9 above)
Comment (sheldon-roberts) [reply flair/models/generative_classification_model.py]: no it isnt, oops!

File: flair/training_utils.py (hunk 11 above)
Comment (sheldon-roberts) [reply flair/training_utils.py]: I did some more looking into the weight tying and realized it wasn't working as I expected. I ended up completely rewriting this function. The good news is that it no longer requires disabling the cache.

**Review by jeffpicard — APPROVED** [review]
This is awesome! I'm sorry I'm not giving a deeper review, but some brief thoughts:
- The constrained decoding is really cool
- The interface for running and instantiating is different than the other classifiers. I didn't read deep enough to understand if this is necessary for an important functionality, but I wonder if most use cases could be served by having `__init__` take a `TransformerEmbeddings` like the other models and then instantiating the causal_model and tokenizer inside the class rather than making the person using the class do it. (I could totally be wrong about this too... just a first impression)
- No unit tests (I know, just a time thing. shouldn't complain about this when the core is so great)

**Comment by sheldon-roberts** [comment]
> * The interface for running and instantiating is different than the other classifiers. I didn't read deep enough to understand if this is necessary for an important functionality, but I wonder if most use cases could be served by having `__init__` take a `TransformerEmbeddings` like the other models and then instantiating the causal_model and tokenizer inside the class rather than making the person using the class do it. (I could totally be wrong about this too... just a first impression)

@jeffpicard This is a good point and I agree that it's not ideal to be so different from the other classifiers. The reason I didn't have this class build the `causal_model` internally is to support weight sharing. If we have multiple generative tasks training with MTL, I want the `causal_model` weights to be shared between each task. So I felt it made more sense to define the `causal_model` externally, and pass it to each task that shares the model. I'm definitely open to ideas if you can think of a cleaner solution.

**Comment by jeffpicard** [comment]
I see. That weight tying looks pretty hairy. I assume you can't tie 3+ things together:
```python
_tie_weights_recursively(a,b)
_tie_weights_recursively(a,c)
# --> b,c are also tied
```
(again, out of my element). My only other thought is a new class of `Embeddings` for like `CausalEmbeddings` or something.

Totally out of scope of this PR. LGTM to merge :)

**Comment by jeffpicard** [comment]
In case you hadn't seen this one before: in the python scripts you pasted in the top comment, if you change the first \`\`\` in the markdown to \`\`\`python    it will highlight the syntax 😄

**Comment by MattGPT-ai** [comment]
Oh I didn't realize, but I think this needs to be merged into `main`

**Comment by sheldon-roberts** [comment]
> Oh I didn't realize, but I think this needs to be merged into `main`

oh shoot. What's the best way to fix this? Should I open a new PR to merge it to main?
